src/networkd.c

```diff
@@ -1460,29 +1474,41 @@ _netplan_netdef_write_networkd(
 gboolean
 _netplan_networkd_write_wait_online(const NetplanState* np_state, const char* rootdir)
 {
+    // Set of all current network interfaces, potentially non yet renamed
+    GHashTable* system_interfaces = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    _netplan_query_system_interfaces(system_interfaces);
+
     // Hash set of non-optional interfaces to wait for
-    GHashTable* non_optional_interfaces = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    GHashTable* non_optional_interfaces = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
     NetplanStateIterator iter;
     netplan_state_iterator_init(np_state, &iter);
     while (netplan_state_iterator_has_next(&iter)) {
-        const NetplanNetDefinition* def = netplan_state_iterator_next(&iter);
+        NetplanNetDefinition* def = netplan_state_iterator_next(&iter);
         if (def->backend != NETPLAN_BACKEND_NETWORKD)
             continue;
+
         /* When activation-mode is used we default to being optional.
          * Otherwise, systemd might wait indefinitely for the interface to
          * come online.
          */
         if (!(def->optional || def->activation_mode)) {
-            if (!netplan_netdef_has_match(def)) { // no matching => single interface
-                g_hash_table_add(non_optional_interfaces, g_strdup(def->id));
-            } else if (def->set_name) { // matching on a single interface
-                g_hash_table_add(non_optional_interfaces, g_strdup(def->set_name));
+            // Check if we have any IP configuration
+            struct address_iter* addr_iter = _netplan_netdef_new_address_iter(def);
+            gboolean any_ips = _netplan_address_iter_next(addr_iter) != NULL || netplan_netdef_get_link_local_ipv4(def) || netplan_netdef_get_link_local_ipv6(def);
```

Comment:
question: should dhcp4 and dhcp6 be considered here? When I have the config below I get `enp3s0:carrier` but according to your last comment it should be `:degraded`?

```yaml
network:
  ethernets:
    enp3s0:
      dhcp4: true
      link-local: []
```

Reply:
Good catch! Of course, it needs to be considered. Fixing...